### PR TITLE
[1LP][RFR] Automate tenant switching of LDAP user

### PR DESCRIPTION
**Pull request by nachandr**
### PRT Run
{{py.test: cfme/tests/configure/test_access_control.py -k 'switch_between_tenants'}}

### CODE REVIEW

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 90ebda1323ca137910b9ae5095670ffeb5776392

The *global* fixture **`openldap_auth_provider`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**
    - *test_tenant_ldap_group_switch_between_tenants*

 The *global* fixture **`setup_openldap_auth_provider`** was changed, but I didn't find where it's used.
Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
Lint report for commit 90ebda1323ca137910b9ae5095670ffeb5776392:

`cfme/tests/configure/test_access_control.py`:
- :red_circle: Line 2003:10: **F821** *undefined name 'test_user'*
- :red_circle: Line 2009:45: **F821** *undefined name 'group_names'*
- :heavy_exclamation_mark: Line 2017:1: **W293** *blank line contains whitespace*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit b26800a10f18e070876d2fb5b9b7a03f98390bc8:

`cfme/tests/configure/test_access_control.py`:
- :bangbang: Line 56:37: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 56:39: **E251** *unexpected spaces around keyword / parameter equals*
- :heavy_exclamation_mark: Line 2029:8: **W291** *trailing whitespace*
- :red_circle: Line 2031:5: **F841** *local variable 'tenant1' is assigned to but never used*
- :red_circle: Line 2032:5: **F841** *local variable 'tenant2' is assigned to but never used*
- :red_circle: Line 2058:9: **F821** *undefined name 'soft_assert'*
- :red_circle: Line 2058:46: **F821** *undefined name 'user_obj'*
- :red_circle: Line 2064:13: **F821** *undefined name 'soft_assert'*
- :red_circle: Line 2070:13: **F821** *undefined name 'soft_assert'*
- :red_circle: Line 2070:40: **F821** *undefined name 'auth_user'*
- :red_circle: Line 2076:13: **F821** *undefined name 'soft_assert'*
- :red_circle: Line 2083:6: **F821** *undefined name 'request'*

Hmmm
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
I detected some fixture changes in commit b26800a10f18e070876d2fb5b9b7a03f98390bc8

 The *global* fixture **`openldap_auth_provider`** was changed, but I didn't find where it's used.
 The *global* fixture **`setup_openldap_auth_provider`** was changed, but I didn't find where it's used.
The *local* fixture **`child_tenant1`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**
    - *test_tenant_ldap_group_switch_between_tenants*

The *local* fixture **`child_tenant2`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**
    - *test_tenant_ldap_group_switch_between_tenants*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
I detected some fixture changes in commit f79519b629b10df4b2e9a6ac6353bd8942a407f1

 The *global* fixture **`openldap_auth_provider`** was changed, but I didn't find where it's used.
The *global* fixture **`setup_openldap_auth_provider`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**

The *local* fixture **`child_tenant1`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**
    - *test_tenant_ldap_group_switch_between_tenants*

The *local* fixture **`child_tenant2`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**
    - *test_tenant_ldap_group_switch_between_tenants*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
I detected some fixture changes in commit dc5c823d54c7ba8e3770245759acea8272be38f7

 The *global* fixture **`openldap_auth_provider`** was changed, but I didn't find where it's used.
The *global* fixture **`setup_openldap_auth_provider`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**

The *local* fixture **`child_tenant1`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**
    - *test_tenant_ldap_group_switch_between_tenants*

The *local* fixture **`child_tenant2`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**
    - *test_tenant_ldap_group_switch_between_tenants*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 6bf571370a877ddd99f95001baefe550f8c0e7cf

 The *global* fixture **`openldap_auth_provider`** was changed, but I didn't find where it's used.
The *global* fixture **`setup_openldap_auth_provider`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**

The *local* fixture **`child_tenant1`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**
    - *test_tenant_ldap_group_switch_between_tenants*

The *local* fixture **`child_tenant2`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**
    - *test_tenant_ldap_group_switch_between_tenants*

The *local* fixture **`setup_openldap_user_group`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 8ccde94ae50a95e1500cec84a0d75d48a51dd74c

 The *global* fixture **`openldap_auth_provider`** was changed, but I didn't find where it's used.
The *global* fixture **`setup_openldap_auth_provider`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**
    - *test_tenant_ldap_group_switch_between_tenants*

The *local* fixture **`two_child_tenants`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**
    - *setup_openldap_user_group*

The *local* fixture **`setup_openldap_user_group`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

File: cfme/tests/configure/test_access_control.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 import fauxfactory
 import pytest
 
+import cfme.utils.auth as authutil
```
[the remark this answers is not in the document]
Comment (nachandr) [reply cfme/tests/configure/test_access_control.py]: Done

File: cfme/tests/configure/test_access_control.py [hunk 2]
```diff
@@ -1977,25 +2004,80 @@ def test_tenant_ldap_group_switch_between_tenants():
...
+                                                   ldap_user.username,
+                                                   group,
+                                                   auth_provider))
+        else:
+            logger.info('All user groups for group switching are evm built-in: {}'
+                    .format(ldap_user.groups))
 
-    """
-    pass
+    user = appliance.collections.users.simple_user(
```
[the remark this answers is not in the document]
Comment (nachandr) [reply cfme/tests/configure/test_access_control.py]: I have moved the code to a fixture since it is a part of test setup .

**Comment by dajoRH** [comment]
I detected some fixture changes in commit de35a80a2b941403ea7a3f6a2414802959317d91

 The *global* fixture **`openldap_auth_provider`** was changed, but I didn't find where it's used.
The *global* fixture **`setup_openldap_auth_provider`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**
    - *test_tenant_ldap_group_switch_between_tenants*

The *local* fixture **`two_child_tenants`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**
    - *setup_openldap_user_group*

The *local* fixture **`setup_openldap_user_group`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

File: cfme/tests/configure/test_access_control.py [hunk 3]
```diff
@@ -50,6 +53,29 @@ def new_role(appliance, name=None):
         vm_restriction='None')
 
 
+def _tenant(appliance, request, name=None):
```
[the remark this answers is not in the document]
Comment (nachandr) [reply cfme/tests/configure/test_access_control.py]: It's not needed anymore. So, I've removed it.

File: cfme/tests/configure/test_access_control.py [hunk 4]
```diff
@@ -1977,25 +2004,80 @@ def test_tenant_ldap_group_switch_between_tenants():
...
+               the active group. After switching to desired group,user is able login
+               via Self Service UI to the desired tenant.
+    """
+    auth_provider = authutil.get_auth_crud('cfme_openldap')
+    ldap_user = authutil.auth_user_data('cfme_openldap', 'uid')[0]
+    retrieved_groups = []
+    for group in ldap_user.groups:
+        if 'evmgroup' not in group.lower():
+            # create group in CFME via retrieve_group which looks it up on auth_provider
+            logger.info(u'Retrieving a user group that is non evm built-in: {}'.format(group))
```
[the remark this answers is not in the document]
Comment (nachandr) [reply cfme/tests/configure/test_access_control.py]: That's an oversight.

File: cfme/tests/configure/test_access_control.py [hunk 5]
```diff
@@ -51,7 +55,59 @@ def new_role(appliance, name=None):
...
+    )
+
+    yield child_tenant1, child_tenant2
+    child_tenant1.delete_if_exists()
+    child_tenant2.delete_if_exists()
+
+
+@pytest.fixture
+def setup_openldap_user_group(appliance, two_child_tenants):
+    auth_provider = get_auth_crud('cfme_openldap')
```
Comment (john-dupuy) [inline cfme/tests/configure/test_access_control.py]: I don't like the call to `get_auth_crud` and `cfme_openldap` twice, I think this should take the fixture `openldap_auth_provider` as an argument.

File: cfme/tests/configure/test_access_control.py [hunk 6]
```diff
@@ -1977,25 +2034,54 @@ def test_tenant_ldap_group_switch_between_tenants():
...
+    user, retrieved_groups = setup_openldap_user_group
+
+    with user:
+        view = navigate_to(appliance.server, 'LoggedIn')
+        # Verify that multiple groups are displayed
+        assert len(view.group_names) > 1, 'Only a single group is displayed for the user'
+        display_other_groups = [g for g in view.group_names if g != view.current_groupname]
+        # Verify that user name is displayed
+        soft_assert(view.current_fullname == user.name,
+                    'user full name "{}" did not match UI display name "{}"'
```
Comment (john-dupuy) [inline cfme/tests/configure/test_access_control.py]: _Optional_ you can use f-strings here

File: cfme/tests/configure/test_access_control.py (hunk 5 above)
Comment (jawatts) [reply cfme/tests/configure/test_access_control.py]: additionally, I would wrap it in a try/except and catch the `KeyError`

Comment (jawatts) [reply cfme/tests/configure/test_access_control.py]: ^ ignore, it is already doing that

File: cfme/tests/configure/test_access_control.py [hunk 7]
```diff
@@ -1977,25 +2034,54 @@ def test_tenant_ldap_group_switch_between_tenants():
...
-            5. Login as mike user who is member of 2 different tenants
-            6. User is able switch between groups - switching is done in a way
-            that current current group which is chosen is writtent into DB as
-            active group. Therefore user who is assigned to more groups must login
-            to Classic UI and switch to desired group. Afterthat he is able login
-            via Self Service UI to desired tenant
-
+            5. Login as 'mike' who is a member of 2 different tenants
+            6. User should be able switch between groups after logging into classis UI.
+               Switching is done in a way that the current group is writtent into DB as
```
Comment (jawatts) [inline cfme/tests/configure/test_access_control.py]: typo here in `writtent`

File: cfme/tests/configure/test_access_control.py (hunk 5 above)
Comment (nachandr) [reply cfme/tests/configure/test_access_control.py]: @john-dupuy , agreed

**Comment by dajoRH** [comment]
I detected some fixture changes in commit e24c8ff6385219ae87ebdd0539319a95a4205a45

The *global* fixture **`openldap_auth_provider`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**
    - *setup_openldap_user_group*

The *global* fixture **`setup_openldap_auth_provider`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**
    - *test_tenant_ldap_group_switch_between_tenants*

The *local* fixture **`two_child_tenants`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**
    - *setup_openldap_user_group*

The *local* fixture **`setup_openldap_user_group`** is used in the following files:
- **cfme/tests/configure/test_access_control.py**

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by jawatts** [comment]
waiting on PRT and travis

**Pull request merged**